### DISCUSSION TITLE
VisionFive 2 ROM addresses for Zephyr RTOS

### DISCUSSION THREAD

**Discussion by kanakshilledar** [zephyrproject-rtos/zephyr#66093] [Q&A]
Hello!

I and @pfarwsi are working on adding support for VisionFive 2 in Zephyr RTOS. I have created a device tree by referencing the [starfive linux](https://github.com/starfive-tech/linux), patches of `hifive_unmatched` [FU740 SoC](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/riscv/sifive/riscv64-fu740.dtsi), and patches of [BeagleV Starlight](https://github.com/zephyrproject-rtos/zephyr/commit/1c584127ba260eb3ab833b1be593d96f8deed122#diff-fcf14c4b7b34fe7a11916195871ae66a59be87a395f28db73e345ebdc828085b).

The porting work is done in this repo: https://github.com/pfarwsi/zephyr-visionfive2/tree/visionfive2

I have created a devicetree for [JH7110 SoC](https://github.com/pfarwsi/zephyr-visionfive2/blob/visionfive2/dts/riscv/starfive/jh7110-visionfive-v2.dtsi).

The build errors out at Linking C executable step while building for `samples/hello_world`.

**Build Instructions:**
```commandline
$ git clone https://github.com/pfarwsi/zephyr-visionfive2.git -b visionfive2
$ west build -p always -b visionfive2_jh7110 samples/hello_world/
```
**Error log:**
```commandline
[91/96] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/kanak/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map
: && ccache /home/kanak/zephyr-sdk-0.16.3/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf zephyr/CMakeFiles/offsets.dir/./arch/riscv/core/offsets/offsets.c.obj -fuse-ld=bfd -T zephyr/linker_zephyr_pre0.cmd -Wl,-Map=/home/kanak/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map -Wl,--whole-archive app/libapp.a zephyr/libzephyr.a zephyr/arch/common/libarch__common.a zephyr/arch/arch/riscv/core/libarch__riscv__core.a zephyr/lib/libc/picolibc/liblib__libc__picolibc.a zephyr/lib/libc/common/liblib__libc__common.a zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a zephyr/drivers/serial/libdrivers__serial.a zephyr/drivers/timer/libdrivers__timer.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a -L"/home/kanak/zephyr-sdk-0.16.3/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.2.0/rv64imac_zicsr_zifencei/lp64/medany" -L/home/kanak/zephyrproject/zephyr/build/zephyr -lgcc zephyr/arch/common/libisr_tables.a -mcmodel=medany -mabi=lp64 -march=rv64imac_zicsr_zifencei -Wl,--gc-sections -Wl,--build-id=none -Wl,--sort-common=descending -Wl,--sort-section=alignment -Wl,-u,_OffsetAbsSyms -Wl,-u,_ConfigAbsSyms -nostdlib -static -Wl,-X -Wl,-N -Wl,--orphan-handling=warn -Wl,-no-pie -DPICOLIBC_LONG_LONG_PRINTF_SCANF --specs=picolibc.specs -lc -lgcc && cd /home/kanak/zephyrproject/zephyr/build/zephyr && /usr/bin/cmake -E true
/home/kanak/zephyr-sdk-0.16.3/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.2.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: invalid origin for memory region ROM
/home/kanak/zephyr-sdk-0.16.3/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.2.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: invalid length for memory region ROM
/home/kanak/zephyr-sdk-0.16.3/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.2.0/../../../../riscv64-zephyr-elf/bin/ld.bfd:zephyr/linker_zephyr_pre0.cmd:44: undefined symbol 'ROM_BASE' referenced in expression
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/kanak/zephyrproject/zephyr/build
```

It gives error of undefined symbol for `ROM_BASE`. 

**Full build log:** https://pastebin.com/JHLZrx66

Where is the memory addresses for the boards defined in Zephyr RTOS?
How can I solve the ROM_BASE issue?

Thanks!
Kanak Shilledar

**Comment by ndrs-pst** [comment]
@kanakshilledar, it seems the issue has been resolved, right?
I've successfully built the `hello_world` sample by integrating your port into the main repository at commit 27b1f4eb7f9d887d8fda81cbdc52f651223e75e9.
I suspect the problem was related to this, and since you've already set `CONFIG_XIP=n`, that might have been the fix.

https://github.com/zephyrproject-rtos/zephyr/blob/6f6fc4e25c3175d5a2587b0e6f3d858b4c336e99/include/zephyr/arch/riscv/common/linker.ld#L39-L60

**Reply by kanakshilledar** [reply]
Yes, the issue has been fixed and the `sample/hello_world` application is getting built successfully. `CONFIG_XIP=n` was missing so adding it in `boards/riscv/visionfive2_jh7110/visionfive2_jh7110_defconfig` fixed the issue.
